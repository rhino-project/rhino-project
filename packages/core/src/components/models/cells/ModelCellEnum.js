import { useMemo } from 'react';
import { useModelAndAttributeFromPath } from '../../../hooks/models';
import { useGlobalComponentForAttribute } from '../../../hooks/overrides';
import CellBadge from '../../table/cells/CellBadge';

const COLOR_PALETTE = ['#0369a1', '#047857', '#6d28d9', '#a21caf', '#be123c'];

export const ModelCellEnumBase = (props) => {
  const { getValue, model, path } = props;
  const { attribute } = useModelAndAttributeFromPath(model, path);

  const style = useMemo(() => {
    if (!getValue()) return {};

    const idx = attribute?.enum?.findIndex((e) => e === getValue());
    const backgroundColor =
      idx >= 0 ? COLOR_PALETTE[idx % COLOR_PALETTE.length] : COLOR_PALETTE[0];

    return { backgroundColor };
  }, [attribute, getValue]);

  return <CellBadge style={style} color="none" {...props} />;
};

const ModelCellEnum = (props) =>
  useGlobalComponentForAttribute('ModelCellEnum', ModelCellEnumBase, props);

export default ModelCellEnum;
