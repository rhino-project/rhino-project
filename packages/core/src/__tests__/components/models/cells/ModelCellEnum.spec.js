import { render } from '@testing-library/react';
import ModelCellEnum from 'rhino/components/models/cells/ModelCellEnum';
import { sharedCellTests } from './sharedCellTests';

describe('ModelCellEnum', () => {
  sharedCellTests(ModelCellEnum);

  it(`adds style as a prop based on matched enum value`, () => {
    const getValue = () => 'bar';

    const { asFragment } = render(
      <ModelCellEnum
        model={{
          properties: {
            dummy: {
              type: 'string',
              enum: ['foo', 'bar']
            }
          }
        }}
        path="dummy"
        getValue={getValue}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it(`uses default style value if match is not found`, () => {
    const getValue = () => 'baz';

    const { asFragment } = render(
      <ModelCellEnum
        model={{
          properties: {
            dummy: {
              type: 'string',
              enum: ['foo', 'bar']
            }
          }
        }}
        path="dummy"
        getValue={getValue}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
