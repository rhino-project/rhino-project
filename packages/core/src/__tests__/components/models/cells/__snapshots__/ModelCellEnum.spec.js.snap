// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelCellEnum adds style as a prop based on matched enum value 1`] = `
<DocumentFragment>
  <span
    class="badge bg-secondary"
    style="background-color: rgb(4, 120, 87);"
  >
    bar
  </span>
</DocumentFragment>
`;

exports[`ModelCellEnum should render the component with className from inherited props 1`] = `
<DocumentFragment>
  <div
    class="dummy-class"
  >
    baz
  </div>
</DocumentFragment>
`;

exports[`ModelCellEnum should render the empty text when value is nullish 1`] = `
<DocumentFragment>
  <div>
    -
  </div>
</DocumentFragment>
`;

exports[`ModelCellEnum should render the overridden empty text when value is nullish 1`] = `
<DocumentFragment>
  <div>
    baz
  </div>
</DocumentFragment>
`;

exports[`ModelCellEnum should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelCellEnum should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelCellEnum should render with global override shorthand for model and attribute 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelCellEnum uses default style value if match is not found 1`] = `
<DocumentFragment>
  <span
    class="badge bg-secondary"
    style="background-color: rgb(3, 105, 161);"
  >
    baz
  </span>
</DocumentFragment>
`;
